File: tests/e2e/java_client/src/test/java/com/example/G3StorageReadAvroTest.java
package com.example;

import com.google.api.gax.core.NoCredentialsProvider;
import com.google.api.gax.grpc.GrpcTransportChannel;
import com.google.api.gax.rpc.FixedTransportChannelProvider;
import com.google.cloud.NoCredentials;
import com.google.cloud.bigquery.BigQuery;
import com.google.cloud.bigquery.BigQueryOptions;
import com.google.cloud.bigquery.DatasetId;
import com.google.cloud.bigquery.DatasetInfo;
import com.google.cloud.bigquery.Field;
import com.google.cloud.bigquery.InsertAllRequest;
import com.google.cloud.bigquery.LegacySQLTypeName;
import com.google.cloud.bigquery.Schema;
import com.google.cloud.bigquery.StandardTableDefinition;
import com.google.cloud.bigquery.TableId;
import com.google.cloud.bigquery.TableInfo;
import com.google.cloud.bigquery.storage.v1.AvroRows;
import com.google.cloud.bigquery.storage.v1.AvroSchema;
import com.google.cloud.bigquery.storage.v1.BigQueryReadClient;
import com.google.cloud.bigquery.storage.v1.BigQueryReadSettings;
import com.google.cloud.bigquery.storage.v1.CreateReadSessionRequest;
import com.google.cloud.bigquery.storage.v1.DataFormat;
import com.google.cloud.bigquery.storage.v1.ReadRowsRequest;
import com.google.cloud.bigquery.storage.v1.ReadRowsResponse;
import com.google.cloud.bigquery.storage.v1.ReadSession;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import org.apache.avro.Schema.Parser;
import org.apache.avro.file.DataFileReader;
import org.apache.avro.file.DataFileWriter;
import org.apache.avro.generic.GenericDatumReader;
import org.apache.avro.generic.GenericDatumWriter;
import org.apache.avro.generic.GenericRecord;
import org.apache.avro.io.BinaryDecoder;
import org.apache.avro.io.DecoderFactory;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * E2E: G3 Storage Read API — Avro wire format (ADR 0030).
 *
 * Three tests, all critical: Avro is Java's default format for the
 * Storage Read client. The default-Avro test catches the case where
 * a Java consumer forgets to set the format and the emulator silently
 * fails. The round-trip-to-disk test is the load-bearing
 * cross-implementation interop proof — it materialises a real .avro
 * file on the container's filesystem via the canonical
 * org.apache.avro DataFileWriter/Reader implementation (NOT fastavro)
 * and verifies the bytes round-trip.
 */
class G3StorageReadAvroTest {
    private static final String REST_URL = System.getenv("BQEMU_REST_URL") != null
            ? System.getenv("BQEMU_REST_URL")
            : "http://localhost:9050";
    private static final String GRPC_ENDPOINT = System.getenv("BQEMU_GRPC_ENDPOINT") != null
            ? System.getenv("BQEMU_GRPC_ENDPOINT")
            : "localhost:9060";
    private static final String PROJECT = "e2e-java-g3";
    private static final String DATASET = "g3_java_ds";
    private static final String TABLE = "avro_rows";

    private static BigQuery bqClient;

    @BeforeAll
    static void setUp() {
        bqClient = BigQueryOptions.newBuilder()
                .setProjectId(PROJECT)
                .setHost(REST_URL)
                .setCredentials(NoCredentials.getInstance())
                .build()
                .getService();
        try {
            bqClient.delete(DatasetId.of(PROJECT, DATASET),
                    BigQuery.DatasetDeleteOption.deleteContents());
        } catch (Exception ignore) {
            // absent
        }
        bqClient.create(DatasetInfo.newBuilder(DATASET).setLocation("US").build());

        Schema schema = Schema.of(
                Field.newBuilder("id", LegacySQLTypeName.INTEGER)
                        .setMode(Field.Mode.REQUIRED).build(),
                Field.newBuilder("name", LegacySQLTypeName.STRING).build(),
                Field.newBuilder("score", LegacySQLTypeName.INTEGER).build());
        bqClient.create(TableInfo.of(
                TableId.of(PROJECT, DATASET, TABLE),
                StandardTableDefinition.of(schema)));

        InsertAllRequest.Builder ins = InsertAllRequest.newBuilder(
                TableId.of(PROJECT, DATASET, TABLE));
        for (Object[] row : new Object[][]{
                {1L, "Alice", 90L},
                {2L, "Bob", 70L},
                {3L, "Carol", 85L},
        }) {
            Map<String, Object> m = new HashMap<>();
            m.put("id", row[0]);
            m.put("name", row[1]);
            m.put("score", row[2]);
            ins.addRow(m);
        }
        bqClient.insertAll(ins.build());
    }

    @AfterAll
    static void tearDown() {
        try {
            bqClient.delete(DatasetId.of(PROJECT, DATASET),
                    BigQuery.DatasetDeleteOption.deleteContents());
        } catch (Exception ignore) {
            // best-effort
        }
    }

    /**
     * Default-Avro path: omit setDataFormat entirely. The Java BQ
     * Storage Read client defaults to AVRO; this test catches the
     * case where the emulator forgets to support that default.
     */
    @Test
    void defaultDataFormatIsAvro() throws Exception {
        ManagedChannel channel = ManagedChannelBuilder.forTarget(GRPC_ENDPOINT)
                .usePlaintext()
                .build();
        try {
            BigQueryReadSettings settings = BigQueryReadSettings.newBuilder()
                    .setCredentialsProvider(NoCredentialsProvider.create())
                    .setTransportChannelProvider(FixedTransportChannelProvider.create(
                            GrpcTransportChannel.create(channel)))
                    .build();
            try (BigQueryReadClient readClient = BigQueryReadClient.create(settings)) {
                // NOTE: NO setDataFormat call — exercises the default
                // path the Java client uses out of the box.
                ReadSession session = readClient.createReadSession(
                        CreateReadSessionRequest.newBuilder()
                                .setParent("projects/" + PROJECT)
                                .setReadSession(ReadSession.newBuilder()
                                        .setTable("projects/" + PROJECT
                                                + "/datasets/" + DATASET
                                                + "/tables/" + TABLE)
                                        .build())
                                .setMaxStreamCount(1)
                                .build());
                // The server may return ARROW for the unspecified
                // default (matching real BigQuery's proto3 behaviour)
                // OR AVRO if the Java client substituted it client-
                // side. Either is acceptable; the test ensures the
                // call doesn't error out.
                assertNotNull(session.getDataFormat());
                assertTrue(session.getStreamsCount() >= 1);

                int total = decodeRowsAnyFormat(readClient, session);
                assertEquals(3, total, "expected 3 rows from default-format ReadRows");
            }
        } finally {
            channel.shutdownNow();
        }
    }

    /**
     * Explicit-Avro path: setDataFormat(DataFormat.AVRO). Decode via
     * the canonical org.apache.avro GenericDatumReader and assert the
     * decoded values match the seed rows.
     */
    @Test
    void explicitAvroDataFormatRoundTrips() throws Exception {
        ManagedChannel channel = ManagedChannelBuilder.forTarget(GRPC_ENDPOINT)
                .usePlaintext()
                .build();
        try {
            BigQueryReadSettings settings = BigQueryReadSettings.newBuilder()
                    .setCredentialsProvider(NoCredentialsProvider.create())
                    .setTransportChannelProvider(FixedTransportChannelProvider.create(
                            GrpcTransportChannel.create(channel)))
                    .build();
            try (BigQueryReadClient readClient = BigQueryReadClient.create(settings)) {
                ReadSession session = readClient.createReadSession(
                        CreateReadSessionRequest.newBuilder()
                                .setParent("projects/" + PROJECT)
                                .setReadSession(ReadSession.newBuilder()
                                        .setTable("projects/" + PROJECT
                                                + "/datasets/" + DATASET
                                                + "/tables/" + TABLE)
                                        .setDataFormat(DataFormat.AVRO)
                                        .build())
                                .setMaxStreamCount(1)
                                .build());
                assertEquals(DataFormat.AVRO, session.getDataFormat());
                AvroSchema avroSchema = session.getAvroSchema();
                assertNotNull(avroSchema);
                assertNotNull(avroSchema.getSchema());

                org.apache.avro.Schema parsed = new Parser().parse(avroSchema.getSchema());
                List<GenericRecord> rows = readAvroRows(readClient, session, parsed);
                assertEquals(3, rows.size());
                Map<Long, GenericRecord> byId = new HashMap<>();
                for (GenericRecord r : rows) {
                    byId.put((Long) r.get("id"), r);
                }
                assertEquals("Alice", byId.get(1L).get("name").toString());
                assertEquals(90L, byId.get(1L).get("score"));
                assertEquals("Carol", byId.get(3L).get("name").toString());
            }
        } finally {
            channel.shutdownNow();
        }
    }

    /**
     * Round-trip-to-disk: read Avro rows from the wire, write them
     * into an .avro Object Container File on disk via the canonical
     * DataFileWriter, re-read with DataFileReader, assert byte-for-
     * byte equality of decoded records.
     *
     * This is the load-bearing cross-implementation interop proof
     * for G3 (ADR 0030 §6). If the emulator ever drifts away from
     * the canonical Apache Avro wire format, this test catches it
     * before any user reports it.
     */
    @Test
    void roundTripToAvroFileOnDisk(@TempDir Path tempDir) throws Exception {
        ManagedChannel channel = ManagedChannelBuilder.forTarget(GRPC_ENDPOINT)
                .usePlaintext()
                .build();
        List<GenericRecord> wireRows;
        org.apache.avro.Schema parsedSchema;
        try {
            BigQueryReadSettings settings = BigQueryReadSettings.newBuilder()
                    .setCredentialsProvider(NoCredentialsProvider.create())
                    .setTransportChannelProvider(FixedTransportChannelProvider.create(
                            GrpcTransportChannel.create(channel)))
                    .build();
            try (BigQueryReadClient readClient = BigQueryReadClient.create(settings)) {
                ReadSession session = readClient.createReadSession(
                        CreateReadSessionRequest.newBuilder()
                                .setParent("projects/" + PROJECT)
                                .setReadSession(ReadSession.newBuilder()
                                        .setTable("projects/" + PROJECT
                                                + "/datasets/" + DATASET
                                                + "/tables/" + TABLE)
                                        .setDataFormat(DataFormat.AVRO)
                                        .build())
                                .setMaxStreamCount(1)
                                .build());
                parsedSchema = new Parser().parse(session.getAvroSchema().getSchema());
                wireRows = readAvroRows(readClient, session, parsedSchema);
            }
        } finally {
            channel.shutdownNow();
        }

        // Materialise into an OCF on disk using the canonical
        // org.apache.avro DataFileWriter.
        File ocfPath = tempDir.resolve("e2e_java_dump.avro").toFile();
        GenericDatumWriter<GenericRecord> writer = new GenericDatumWriter<>(parsedSchema);
        try (DataFileWriter<GenericRecord> fileWriter = new DataFileWriter<>(writer)) {
            fileWriter.create(parsedSchema, ocfPath);
            for (GenericRecord row : wireRows) {
                fileWriter.append(row);
            }
        }

        // Verify the file has the canonical OCF magic bytes.
        byte[] header = new byte[4];
        try (java.io.FileInputStream fis = new java.io.FileInputStream(ocfPath)) {
            int read = fis.read(header);
            assertEquals(4, read);
            assertEquals('O', (char) header[0]);
            assertEquals('b', (char) header[1]);
            assertEquals('j', (char) header[2]);
            assertEquals(1, header[3]);
        }

        // Re-read via DataFileReader and assert the decoded rows
        // equal the wire rows. Proves the canonical Java Avro
        // implementation accepts the bytes the emulator emitted.
        GenericDatumReader<GenericRecord> datumReader = new GenericDatumReader<>(parsedSchema);
        try (DataFileReader<GenericRecord> fileReader = new DataFileReader<>(ocfPath, datumReader)) {
            List<GenericRecord> decoded = new ArrayList<>();
            while (fileReader.hasNext()) {
                decoded.add(fileReader.next());
            }
            assertEquals(wireRows.size(), decoded.size());
            for (int i = 0; i < wireRows.size(); i++) {
                assertEquals(
                        wireRows.get(i).get("id"),
                        decoded.get(i).get("id"));
                assertEquals(
                        wireRows.get(i).get("name").toString(),
                        decoded.get(i).get("name").toString());
            }
        }
    }

    /** Decode one ReadSession's worth of AvroRows messages → GenericRecords. */
    private static List<GenericRecord> readAvroRows(
            BigQueryReadClient client,
            ReadSession session,
            org.apache.avro.Schema schema) {
        List<GenericRecord> out = new ArrayList<>();
        GenericDatumReader<GenericRecord> reader = new GenericDatumReader<>(schema);
        for (int i = 0; i < session.getStreamsCount(); i++) {
            String streamName = session.getStreams(i).getName();
            for (ReadRowsResponse resp : client.readRowsCallable().call(
                    ReadRowsRequest.newBuilder()
                            .setReadStream(streamName)
                            .build())) {
                AvroRows avroRows = resp.getAvroRows();
                if (avroRows == null || avroRows.getSerializedBinaryRows().isEmpty()) {
                    continue;
                }
                try {
                    BinaryDecoder decoder = DecoderFactory.get().binaryDecoder(
                            new ByteArrayInputStream(
                                    avroRows.getSerializedBinaryRows().toByteArray()),
                            null);
                    for (long r = 0; r < resp.getRowCount(); r++) {
                        out.add(reader.read(null, decoder));
                    }
                } catch (Exception exc) {
                    throw new RuntimeException(
                            "Avro decode failed: " + exc.getMessage(), exc);
                }
            }
        }
        return out;
    }

    /**
     * Decode rows from either Arrow or Avro session — used by the
     * default-format test which doesn't know upfront which the
     * server will pick.
     */
    private static int decodeRowsAnyFormat(
            BigQueryReadClient client,
            ReadSession session) throws Exception {
        if (session.getDataFormat() == DataFormat.AVRO) {
            org.apache.avro.Schema parsed = new Parser().parse(
                    session.getAvroSchema().getSchema());
            return readAvroRows(client, session, parsed).size();
        }
        // Arrow path — minimal count, schema decoding handled in StorageReadTest.
        int total = 0;
        for (int i = 0; i < session.getStreamsCount(); i++) {
            String streamName = session.getStreams(i).getName();
            for (ReadRowsResponse resp : client.readRowsCallable().call(
                    ReadRowsRequest.newBuilder()
                            .setReadStream(streamName)
                            .build())) {
                total += (int) resp.getRowCount();
            }
        }
        return total;
    }
}

File: tests/e2e/go_client/g3_storage_read_avro_test.go
// G3 Storage Read API E2E for the bqemulator — Avro wire format
// using the Go BigQuery Storage Read gRPC client + linkedin/goavro
// for decoding. Mirrors storage_read_test.go's Arrow shape but requests
// AVRO and asserts at the decoded-row level so a "bytes that
// proto-validate but no Avro decoder accepts" regression is caught.

package e2e

import (
	"context"
	"io"
	"testing"
	"time"

	"cloud.google.com/go/bigquery"
	storage "cloud.google.com/go/bigquery/storage/apiv1"
	storagepb "cloud.google.com/go/bigquery/storage/apiv1/storagepb"
	"github.com/linkedin/goavro/v2"
	"google.golang.org/api/option"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

const g3Project = "e2e-go-g3"

type avroRow struct {
	ID    int64  `bigquery:"id"`
	Name  string `bigquery:"name"`
	Score int64  `bigquery:"score"`
}

func setupAvroReadTable(t *testing.T, ctx context.Context) func() {
	t.Helper()
	client, err := bigquery.NewClient(
		ctx,
		g3Project,
		option.WithEndpoint(bqAPIBase()),
		option.WithoutAuthentication(),
	)
	if err != nil {
		t.Fatalf("rest client: %v", err)
	}
	datasetID := "g3_go_ds"
	tableID := "avro_rows"
	ds := client.Dataset(datasetID)
	_ = ds.DeleteWithContents(ctx)
	if err := ds.Create(ctx, &bigquery.DatasetMetadata{Location: "US"}); err != nil {
		t.Fatalf("dataset create: %v", err)
	}
	schema := bigquery.Schema{
		{Name: "id", Type: bigquery.IntegerFieldType, Required: true},
		{Name: "name", Type: bigquery.StringFieldType},
		{Name: "score", Type: bigquery.IntegerFieldType},
	}
	if err := ds.Table(tableID).Create(ctx, &bigquery.TableMetadata{Schema: schema}); err != nil {
		t.Fatalf("table create: %v", err)
	}
	if err := ds.Table(tableID).Inserter().Put(ctx, []avroRow{
		{ID: 1, Name: "Alice", Score: 90},
		{ID: 2, Name: "Bob", Score: 70},
		{ID: 3, Name: "Carol", Score: 85},
	}); err != nil {
		t.Fatalf("insert: %v", err)
	}
	return func() {
		_ = ds.DeleteWithContents(ctx)
		_ = client.Close()
	}
}

// TestG3StorageReadAvroDecodesViaGoavro exercises the Avro wire path
// from Go: CreateReadSession + ReadRows with DataFormat_AVRO, decode
// every row via linkedin/goavro, and assert the decoded values equal
// the seeded rows.
func TestG3StorageReadAvroDecodesViaGoavro(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
	defer cancel()
	teardown := setupAvroReadTable(t, ctx)
	defer teardown()

	conn, err := grpc.NewClient(
		grpcEndpoint(),
		grpc.WithTransportCredentials(insecure.NewCredentials()),
	)
	if err != nil {
		t.Fatalf("grpc.NewClient: %v", err)
	}
	defer conn.Close()

	readClient, err := storage.NewBigQueryReadClient(
		ctx,
		option.WithGRPCConn(conn),
		option.WithoutAuthentication(),
	)
	if err != nil {
		t.Fatalf("NewBigQueryReadClient: %v", err)
	}
	defer readClient.Close()

	session, err := readClient.CreateReadSession(ctx, &storagepb.CreateReadSessionRequest{
		Parent: "projects/" + g3Project,
		ReadSession: &storagepb.ReadSession{
			Table:      "projects/" + g3Project + "/datasets/g3_go_ds/tables/avro_rows",
			DataFormat: storagepb.DataFormat_AVRO,
		},
		MaxStreamCount: 1,
	})
	if err != nil {
		t.Fatalf("CreateReadSession: %v", err)
	}
	if len(session.GetStreams()) == 0 {
		t.Fatal("CreateReadSession returned no streams")
	}
	if session.GetDataFormat() != storagepb.DataFormat_AVRO {
		t.Fatalf("session data_format = %v, want AVRO", session.GetDataFormat())
	}
	schemaJSON := session.GetAvroSchema().GetSchema()
	if schemaJSON == "" {
		t.Fatal("session missing avro_schema")
	}

	codec, err := goavro.NewCodec(schemaJSON)
	if err != nil {
		t.Fatalf("goavro.NewCodec: %v", err)
	}

	type row map[string]any
	var decoded []row
	for _, stream := range session.GetStreams() {
		rows, err := readClient.ReadRows(ctx, &storagepb.ReadRowsRequest{ReadStream: stream.GetName()})
		if err != nil {
			t.Fatalf("ReadRows: %v", err)
		}
		for {
			resp, err := rows.Recv()
			if err == io.EOF {
				break
			}
			if err != nil {
				t.Fatalf("ReadRows.Recv: %v", err)
			}
			payload := resp.GetAvroRows().GetSerializedBinaryRows()
			if len(payload) == 0 {
				continue
			}
			// Naked rows — linearly decode resp.GetRowCount() records
			// back-to-back, using goavro's (value, remainder) return
			// shape to track the cursor.
			cursor := payload
			for j := int64(0); j < resp.GetRowCount(); j++ {
				native, remaining, err := codec.NativeFromBinary(cursor)
				if err != nil {
					t.Fatalf("goavro decode at row %d: %v", j, err)
				}
				m, ok := native.(map[string]any)
				if !ok {
					t.Fatalf("expected map row, got %T", native)
				}
				decoded = append(decoded, m)
				cursor = remaining
			}
		}
	}

	if len(decoded) != 3 {
		t.Fatalf("expected 3 decoded Avro rows, got %d", len(decoded))
	}
	// goavro decodes nullable-union values as map[string]any{"<type>": value}
	// (so we walk one level into the union wrapper for non-REQUIRED columns
	// like ``name`` / ``score``) and bare values for REQUIRED columns like
	// ``id`` (the emulator emits a bare ``T`` for ``mode='REQUIRED'`` per
	// real BigQuery's documented Avro shape).
	getString := func(m row, key string) string {
		if wrapper, ok := m[key].(map[string]any); ok {
			if s, ok := wrapper["string"].(string); ok {
				return s
			}
		}
		if s, ok := m[key].(string); ok {
			return s
		}
		return ""
	}
	byID := map[int64]row{}
	for _, r := range decoded {
		id, _ := r["id"].(int64)
		byID[id] = r
	}
	if got := getString(byID[1], "name"); got != "Alice" {
		t.Errorf("row 1 name = %q, want Alice", got)
	}
	if got := getString(byID[3], "name"); got != "Carol" {
		t.Errorf("row 3 name = %q, want Carol", got)
	}
}
